src/Dict/Helpers.elm

```diff
@@ -0,0 +1,37 @@
+
+module Dict.Helpers exposing ( fromListBy, fromListAppendBy )
+
+import Dict exposing ( Dict )
+
+fromListBy : (a -> comparable) -> List a -> Dict comparable a
+fromListBy getKey xs =
+  let initState = Dict.empty
+      action x state =
+        Dict.insert (getKey x) x state
+  in List.foldl action initState xs
+
+-- Given for example a list of records (type a)
+--   xs : List a
+-- each of which has a field id : Maybe Int,
+-- then
+--   fromListAppendBy .id xs
+-- computes a dictionary whose keys are the ids and values are lists of all records with that id.
+-- 
+-- Note that this conversion is linear (assuming that getKey is constant complexity) in the length of the list xs.
```

Comment:
> is linear in the length

Meaning linear in complexity, as in `O(n)`? Just checking I understand the comment correctly.

Reply:
Also, sounds like [`Dict.Extra.groupBy`](https://package.elm-lang.org/packages/elm-community/dict-extra/latest/Dict-Extra#groupBy)

Reply:
Damn, `Dict.Extra.groupBy` is exactly what I was looking for. And I have no idea how I could have missed it since it's the first function in the module. I actually found `Dict.Extra.fromListBy`, but then reimplemented it myself - I didn't want to include the whole module because of one function for some reason.

Reply:
Yes, `O(n)`, the worst case asymptotic complexity.

Reply:
Wait, this is more like O(n * log(n))